xdsl/traits.py

```diff
@@ -68,6 +68,40 @@ def verify(self, op: Operation) -> None:
         raise VerifyException(f"'{op.name}' expects parent op to be one of {names}")
 
 
+@dataclass(frozen=True)
+class HasAncestor(OpTrait):
+    """
+    Constraint the operation to have a specific operation as ancestor, i.e. transitive
+    parent.
+    """
+
+    parameters: tuple[type[Operation], ...]
+
+    def __init__(self, *parameters: type[Operation]):
+        if not parameters:
+            raise ValueError("parameters must not be empty")
+        super().__init__(parameters)
+
+    def verify(self, op: Operation) -> None:
+        if self.get_ancestor(op) is None:
+            if len(self.parameters) == 1:
+                raise VerifyException(
+                    f"'{op.name}' expects ancestor op '{self.parameters[0].name}'"
+                )
+            names = ", ".join(f"'{p.name}'" for p in self.parameters)
+            raise VerifyException(
+                f"'{op.name}' expects ancestor op to be one of {names}"
+            )
+
+    def get_ancestor(self, op: Operation) -> Operation | None:
+        parent = op.parent_op()
+        while not isinstance(parent, self.parameters):
+            if parent is None:
+                return None
+            parent = parent.parent_op()
+        return parent
```

Comment:
Should this be a helper on operation? It has a very similar vibe to `walk` and `walk_backwards`, I think a `walk_ancestors` would make sense. To minimise the scope creep, I'd like to propose splitting the iteration from the predicate first, putting both helpers here, and then eventually migrate them.
```suggestion
    def walk_ancestors(self, op: Operation) -> Iterator[Operation]:
        """Iterates over the ancestors of an operation, including the input"""
        curr = op
        yield curr
        while (curr := curr.parent_op()) is not None:
            yield curr

    def get_ancestor(self, op: Operation) -> Operation | None:
        ancestors = self.walk_ancestors(op)
        matching_ancestors = (a for a in ancestors if isinstance(a, self.parameters))
        return next(matching_ancestors, None)
```

Reply:
Why not!

Reply:
Will do but if anyone else disagrees and we start bikeshedding on core, I'm reversing to add functionality first :+1:

Reply:
Oh wait; didn't see your idea of splitting but still in traits. Perfect on the concern ahah